File: CPP/Palindrome.cpp
#include <iostream>

using namespace std;

bool isPalindrome(string word) {
    int left = 0;
    int right = word.length() - 1;
    while (left < right) {
        if (word[left] != word[right])
            return false;
        left++;
        right--;
    }
    return true;
}

bool isPalindromeRecursive(string word, int left, int right) {
    if (left >= right)
        return true;
    if (word[left] != word[right])
        return false;
    return isPalindromeRecursive(word, left + 1, right - 1);
}

int main() {
	cout << isPalindrome("radar") << endl;
	cout << isPalindrome("racecar") << endl;
	cout << isPalindrome("rotator") << endl;

	cout << isPalindromeRecursive("radar", 0, 4) << endl;
	cout << isPalindromeRecursive("racecar", 0, 6) << endl;
	cout << isPalindromeRecursive("rotator", 0, 6) << endl;
}

File: Python/Palindrome.py
def is_palindrome(word: str) -> bool:
    return word == word[::-1]

print(is_palindrome("radar"))
print(is_palindrome("racecar"))
print(is_palindrome("rotator"))

def is_palindrome_iterative(word: str) -> bool:
    start = 0
    end = len(word) - 1
    while start < end:
        if word[start] != word[end]:
            return False
        start += 1
        end -= 1
    return True

print(is_palindrome_iterative("radar"))
print(is_palindrome_iterative("racecar"))
print(is_palindrome_iterative("rotator"))

def is_palindrome_recursive(word: str, start: int, end: int) -> bool:
    if start >= end:
        return True
    if word[start] != word[end]:
        return False
    return is_palindrome_recursive(word, start + 1, end - 1)

print(is_palindrome_recursive("radar", 0, len("radar") - 1))
print(is_palindrome_recursive("racecar", 0, len("racecar") - 1))
print(is_palindrome_recursive("rotator", 0, len("rotator") - 1))

def is_palindrome_recursive_simple(word: str) -> bool:
    if len(word) <= 1:
        return True
    if word[0] != word[-1]:
        return False
    return is_palindrome_recursive_simple(word[1:-1])

print(is_palindrome_recursive_simple("radar"))
print(is_palindrome_recursive_simple("racecar"))
print(is_palindrome_recursive_simple("rotator"))
